File: c/isogram_checker.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>

/**
 * Checks if a given string is an isogram.
 * An isogram is a word or phrase with no repeating letters.
 * 
 * @param str The string to check
 * @return true if the string is an isogram, false otherwise
 */
bool is_isogram(const char *str) {
    // Boolean array to track which letters have been seen
    // 26 letters in the alphabet
    bool letter_seen[26] = {false};
    
    for (int i = 0; str[i] != '\0'; i++) {
        // Convert to lowercase for case-insensitive checking
        char c = tolower(str[i]);
        
        // Only check alphabetic characters
        if (c >= 'a' && c <= 'z') {
            int index = c - 'a'; // Map 'a' to 0, 'b' to 1, etc.
            
            // If we've seen this letter before, it's not an isogram
            if (letter_seen[index]) {
                return false;
            }
            
            // Mark this letter as seen
            letter_seen[index] = true;
        }
    }
    
    return true;
}

int main() {
    // Test cases
    const char *test_words[] = {
        "Dermatoglyphics",
        "isogram",
        "moose",
        "aba",
        "thumbscrew",
        "alphabet",
        ""
    };
    
    int num_tests = sizeof(test_words) / sizeof(test_words[0]);
    
    printf("Isogram Checker using Boolean Array\n\n");
    
    for (int i = 0; i < num_tests; i++) {
        bool result = is_isogram(test_words[i]);
        printf("\"%s\" is %s\n", test_words[i], 
               result ? "an isogram" : "NOT an isogram");
    }
    
    return 0;
}

File: python/isogram_checker.py
def is_isogram(string):
    """
    Checks if a given string is an isogram.
    An isogram is a word or phrase with no repeating letters.
    
    Args:
        string: The string to check
    
    Returns:
        True if the string is an isogram, False otherwise
    """
    # Convert to lowercase to handle case-insensitive checking
    string = string.lower()
    
    # Boolean array to track which letters have been seen
    # 26 letters in the alphabet
    letter_seen = [False] * 26
    
    for char in string:
        # Only check alphabetic characters
        if char.isalpha():
            # Map 'a' to 0, 'b' to 1, etc.
            index = ord(char) - ord('a')
            
            # If we've seen this letter before, it's not an isogram
            if letter_seen[index]:
                return False
            
            # Mark this letter as seen
            letter_seen[index] = True
    
    return True


if __name__ == "__main__":
    # Test cases
    test_words = [
        "Dermatoglyphics",
        "isogram",
        "moose",
        "aba",
        "thumbscrew",
        "alphabet",
        ""
    ]
    
    print("Isogram Checker using Boolean Array\n")
    
    for word in test_words:
        result = is_isogram(word)
        print(f'"{word}" is {"an isogram" if result else "NOT an isogram"}')
